### PR TITLE
HIVE-26045: Detect timed out connections for providers and auto-reconnect

### PR DESCRIPTION
**Pull request by zhangbutao**
### Why are the changes needed?

Check current jdbc connection status(isClosed()) before opening a new connection. 

Add jdbc  properties when opening a jdbc connection. eg connection timeout and retry times.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Qtest

### CODE REVIEW

**Comment by nrg4878** [comment]
I had tested this change locally and it seemed to have worked. I was able to test it after keeping beeline session idle for a while before trying another command.
diff --git a/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java b/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java
index d392a2567a..8d161f72ae 100644
--- a/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java
+++ b/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java
@@ -116,6 +116,7 @@ public AbstractJDBCConnectorProvider(String dbName, DataConnector dataConn, Stri
   }
 
   protected Connection getConnection() {
+    Connection conn;
     try {
       if (!isOpen)
         open();
@@ -123,8 +124,29 @@ protected Connection getConnection() {
       throw new RuntimeException(ce.getMessage(), ce);
     }
 
-    if (handle instanceof Connection)
-      return (Connection)handle;
+    if (handle instanceof Connection) {
+      conn = (Connection) handle;
+      try {
+        if (conn.isValid(3))
+          return conn;
+      } catch (SQLException sqle) {
+        // not error, just re-open the connection
+        LOG.warn("Connection validation failed, resetting connection", sqle);
+      }
+
+      try {
+        close();
+      } catch (Exception e) {
+        LOG.warn("Connection close failed, could have timed out. Resetting", e);
+      }
+
+      try {
+        open();
+        return (Connection) handle;
+      } catch (ConnectException ce) {
+        throw new RuntimeException(ce.getMessage(), ce);
+      }
+    }
 
     throw new RuntimeException("unexpected type for connection handle");
   }
@@ -136,6 +158,8 @@ protected Connection getConnection() {
       } catch (SQLException sqle) {
         LOG.warn("Could not close jdbc connection to " + jdbcUrl, sqle);
         throw new RuntimeException(sqle);
+      } finally {
+        isOpen = false;
       }
     }
   }

Let me know what you think. logically, it is similar to ur implementation. I am using the boolean isOpen while you are using a method.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 1]
```diff
@@ -129,8 +129,20 @@ protected Connection getConnection() {
     throw new RuntimeException("unexpected type for connection handle");
   }
 
+  protected boolean isOpen() {
+    if (handle == null) {
+      return false;
+    }
+    try {
+      return ((Connection)handle).isClosed();
```
[the remark this answers is not in the document]
Comment (nrg4878) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:137]: also isValid() might be better than isClosed(). I dont know if isClosed() returns true if close() has not been called on the connection. isValid() seems to test the health of the connection

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 2]
```diff
@@ -129,8 +129,20 @@ protected Connection getConnection() {
...
+    } catch (SQLException e) {
+      LOG.warn("Could not validate jdbc connection to"+ jdbcUrl, e);
```
Comment (nrg4878) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:139]: nit: need a space after "to" and before the jdbcUrl
so "to " + jdbcUrl

**Comment by zhangbutao** [comment]
> Let me know what you think. logically, it is similar to ur implementation. I am using the boolean isOpen while you are using a method.

@nrg4878 Yes, I wrapped a method to check if conn are closed, and you check conn is alive by isValid().
I think the main difference is that you changes focus on in _AbstractJDBCConnectorProvider::getConnection()_  , so it is suitable for all jdbc datasources(MySQL, Oracle..), but in this way, how can we customize timeouts and retries for each jdbc datasource? meanwhile, each jdbc datasouce may have different timeout parameter or others, it seems that we can not customize jdbc-url for all jdbc datasources in this way.

My change focus on customizing jdbc-url for each jdbc datasource, in this change, i added parameters (connectTimeout and autoReconnect and maxReconnects) in _MySQLConnectorProvider::getConnectionProperties()_, and this parameters can be override by connector DCPROPERTIES. We also can add timeout conn for other jdbc datasource in this way, e.g Oracle.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MySQLConnectorProvider.java [hunk 3]
```diff
@@ -27,14 +27,27 @@
...
 import java.sql.Statement;
+import java.util.Properties;
 
 public class MySQLConnectorProvider extends AbstractJDBCConnectorProvider {
   private static Logger LOG = LoggerFactory.getLogger(MySQLConnectorProvider.class);
 
   private static final String DRIVER_CLASS = "com.mysql.jdbc.Driver";
 
+  public static final String JDBC_MYSQL_CONFIG_PREFIX = "hive.sql.mysql";
+  public static final String JDBC_MYSQL_AUTO_RECONNECT = JDBC_MYSQL_CONFIG_PREFIX + ".auto.reconnect";
```
[the remark this answers is not in the document]
Comment (zhangbutao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/MySQLConnectorProvider.java:38]: These parameters are used in MySQL jdbc-url, and can be override by MySQL connector DCPROPERTIES.  You can check the changer in dataconnector_mysql.q.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 4]
```diff
@@ -129,8 +129,20 @@ protected Connection getConnection() {
...
+    }
+    return false;
+  }
+
   @Override public void close() {
-    if (isOpen) {
```
[the remark this answers is not in the document]
Comment (zhangbutao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:133]: Yes, after this change, this boolen is dead. Will remove it.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java [hunk 5]
```diff
@@ -107,8 +108,7 @@ public AbstractJDBCConnectorProvider(String dbName, DataConnector dataConn, Stri
 
   @Override public void open() throws ConnectException {
     try {
-      handle = DriverManager.getConnection(jdbcUrl, username, password);
-      isOpen = true;
+      handle = DriverManager.getDriver(jdbcUrl).connect(jdbcUrl, getConnectionProperties());
```
[the remark this answers is not in the document]
Comment (zhangbutao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:111]: _DriverManager.getDriver().connect()_ can pass more jdbc parameters(timeout & retrye) when building connection.  I use _MySQLConnectorProvider::getConnectionProperties()_ to pass MySQL connecton timeout and retry..

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java (hunk 2 above)
Comment (zhangbutao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:139]: Thx, will fix it.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangbutao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/jdbc/AbstractJDBCConnectorProvider.java:137]: > also isValid() might be better than isClosed(). I dont know if isClosed() returns true if close() has not been called on the connection. isValid() seems to test the health of the connection

I agree :)

**Comment by nrg4878** [comment]
@zhangbutao will you be able to refresh this patch based on feedback?

**Comment by zhangbutao** [comment]
>  will you be able to refresh this patch based on feedback?

@nrg4878 I can not reopen this closed pr. Should i file a new PR for this ticket?

**Comment by nrg4878** [comment]
That would be great. Thank you

**Comment by zhangbutao** [comment]
Superseded by https://github.com/apache/hive/pull/3595
cc @nrg4878

**Pull request closed without merging**